DietPi-PreBoot | Separate first boot and regular boot steps

FredericGuilbault opened the pull request:
https://github.com/MichaIng/DietPi/issues/2791

This PR separate the pre-network part of the first boot process into it's own .service file. 

( post-login part is yet to be done. )

Tested on RPI   , Qemu-ARM and Qemu-i686

FredericGuilbault committed:
put Apply_DietPi_FirstRun_Settings in it's own file and create a service

FredericGuilbault committed:
Add install section to the service file. It's needed to enable it.

FredericGuilbault committed:
remove Duplication of swap file creation

Fourdee commented:
@FredericGuilbault 

Very nice :+1: 

Many thanks for the PR, i'll take a look and test.

Fourdee commented on dietpi/dietpi-firstboot:
```diff
@@ -0,0 +1,215 @@
+	# Import DietPi-Globals --------------------------------------------------------------
+	/DietPi/dietpi/func/dietpi-obtain_hw_model # Runs every boot to allow e.g. switching SDcards between devices and to be failsafe
+	. /DietPi/dietpi/func/dietpi-globals
+	G_PROGRAM_NAME='DietPi-PreBoot'
```
DietPi-FirstBoot

FredericGuilbault replied:
fixed in 2cf3fad79ff1a3f7805c94d7d9f071661b5075a3

Fourdee commented on rootfs/etc/systemd/system/dietpi-firstboot.service:
```diff
@@ -0,0 +1,14 @@
+[Unit]
+Description=DietPi-First_boot_installation_process
+DefaultDependencies=false
+After= dietpi-preboot.service
```
Extra space needs removing (consistency)

FredericGuilbault replied:
fixed in 2cf3fad79ff1a3f7805c94d7d9f071661b5075a3

Fourdee commented on rootfs/etc/systemd/system/dietpi-firstboot.service:
```diff
@@ -0,0 +1,14 @@
+Type=oneshot
+StandardOutput=tty
+ExecStartPre=
+ExecStart=/bin/dash -c '/DietPi/dietpi/dietpi-firstboot 2>&1 | tee /var/tmp/dietpi/logs/first_boot_install.log'
```
Once run, we could actually remove ```/DietPi/dietpi/dietpi-firstboot``` from system to save 4K+ in RAMdisk, however, would be restored after an ```dietpi-update```.

Would need to add an exclude for ```/DietPi/dietpi/dietpi-firstboot``` in ```dietpi-update```

Fourdee replied:
@FredericGuilbault done with https://github.com/FredericGuilbault/DietPi/commit/c0aa1e8cc044b8328f93bd53fffba6515e82769e

FredericGuilbault replied:
Saving save 4K+ in RAMdisk is a strong argument. 

But If I delete it. there no way to run it twice and I think re-running the install process is part of @MichaIng dev workflow. So IDK how he feel about that. 

Is it why the partition resize script is in `/var/lib/dietpi/services/fs_partition_resize.sh` and not `/DietPi/dietpi/` ?

MichaIng replied:
@FredericGuilbault @Fourdee 
What you think about moving this script to `/var/lib/dietpi/services`? Makes sense to ban code from RAMdisk that only runs once.
> Is it why the partition resize script is in /var/lib/dietpi/services/fs_partition_resize.sh and not /DietPi/dietpi/ ?

Ah yeah exactly, this is the best example and totally makes sense to have both scripts there since both are usually only required a single time.

> But If I delete it. there no way to run it twice and I think re-running the install process is part of @MichaIng dev workflow. So IDK how he feel about that.

Nope it usually is not needed anymore. We never actively call this a second time. But it's good to at least have the chance to do so, e.g. if the first execution failed for some reason. So keep it on the disk, but not in RAM sounds best. And also for testing. E.g. I just wanted to run some speed test of the script with first run setup code inside and outside, if there is a measurable difference. (If not, the cleaner code and less code in RAM is reason enough to separate them)

FredericGuilbault replied:
Moving to `/var/lib/dietpi/service` seem good to me. it solve all the needs and make it consistent with `fs_partition_resize.sh` who have the same role.

FredericGuilbault replied:
Fixed in 7760a2d

Fourdee commented on dietpi/dietpi-firstboot:
```diff
@@ -0,0 +1,215 @@
+	#////////////////////////////////////
+	#
+	# Info:
+	# - filename /DietPi/dietpi/dietpi-frostboot
```
```/DietPi/dietpi/dietpi-firstboot```

Fourdee commented on rootfs/etc/systemd/system/dietpi-firstboot.service:
```diff
@@ -0,0 +1,14 @@
+[Unit]
+Description=DietPi-First_boot_installation_process
```
Spaces are supported in the description. To match our current systemd descriptions, would prefer to see spaces used here aswell.

FredericGuilbault replied:
fixed in 2cf3fad79ff1a3f7805c94d7d9f071661b5075a3

Fourdee reviewed (changes requested):
Nice work :+1: 

Some minor changes required from my end, mostly for consistency.

Fourdee committed:
Remove dietpi-firstboot during updates.

https://github.com/MichaIng/DietPi/pull/2833#pullrequestreview-239478839

FredericGuilbault commented:
Im glad to have your feedback. Im on a contract right now but Ill look at it in the next weeks

MichaIng commented on dietpi/boot:
```diff
@@ -108,9 +108,6 @@
 	if (( $G_DIETPI_INSTALL_STAGE == 10 )); then
 
-		# - Set swap
-		/DietPi/dietpi/func/dietpi-set_swapfile $(grep -m1 '^[[:blank:]]*AUTO_SETUP_SWAPFILE_SIZE=' /DietPi/dietpi.txt | sed 's/^[^=]*=//') "$(grep -m1 '^[[:blank:]]*AUTO_SETUP_SWAPFILE_LOCATION=' /DietPi/dietpi.txt | sed 's/^[^=]*=//')"
```
This needs to stay. First run setup does not run on pre-installed images. But the swapfile or course we do not want to ship with the image file/download, so it needs to be re-created.

But @Fourdee will know more about this, since I never created pre-installed images. Perhaps for those it still makes sense to run all first run setup steps, just have related software installed already, which should not be affected by first run setup steps. Only thing is that users might be confused about the dietpi-software prompt.

Fourdee replied:
Yep needs to stay as it was, for pre-made images. Currently only Allo images, but still needs to be there for all devices.

FredericGuilbault replied:
oh! So far Lysmarine will setup much like Allo. I guess im gonna need it too... lol
I didin't tough about this use case. 

There is to 2 call of  `dietpi-set_swapfile` 
One on `dietpi/boot` and one inside the function `Apply_DietPi_FirstRun_Settings()`. 

**My elements of concern are :** 

I wish that one could be enough. 

I have a strong feeling that once compressed a pre-installed image don't take more space with a swap file in it. 

But we can't assume there will be enough space on the User drive and there is no way to predict this. 

It's over kill to check if we should create a swapfile at every boot.

My hypothesis is either it's not needed in `Apply_DietPi_FirstRun_Settings()` or it should be seperated in it's own one-time service. like `dietpi-fs_partition_resize.service` and `dietpi-firstboot.service`

MichaIng replied:
@FredericGuilbault 
Jep, `Apply_DietPi_FirstRun_Settings()` is not called when `(( $G_DIETPI_INSTALL_STAGE == 10 ))` (pre-installed image) and the other way round. So both need to create the default swap file.

> I have a strong feeling that once compressed a pre-installed image don't take more space with a swap file in it.

Not sure if the swap file is emptied on shutdown, but I guess not, instead it is cleared on boot when initialised most likely. But the main reason against pre-creating it:
- Some SBCs can have multiple RAM size options, e.g. RPi with 256/512/1024M versions and NanoPi M4 at least can have different RAM sizes as well, not to forget native PCs and virtual machines.
- As we by default aim to reach min. 2G overall RAM, this then requires different or even no swap file.

> I wish that one could be enough.

As said, actually the firstrun setup steps should not hurt on pre-installed images as well. Even with Allo/Amiberry/Lysmarine files and services installed, it makes sense to apply locale, timezone, keyboard layout and especially do security relevant things like SSH host key regeneration and machine ID reset.
And network setup according to dietp.txt choices.
Actually I don't see any step that could/should affect a pre-installed software.

@Fourdee 
Not sure in case of Allo, but perhaps we can or even should remove pre-installed images special handling? As said only the `dietpi-software` prompt on first login could confuse or be skipped when having the desired set pre-installed. Although it does not hurt to offer installing additional software on top.
But if wanted, we could simply solve this via `/DietPi/dietpi/.skip_firstrun_installs` flag file which makes DietPi-Login skip dietpi-software execution on first login. What you think?

FredericGuilbault replied:
> As said, actually the firstrun setup steps should not hurt on pre-installed images as well. Even with Allo/Amiberry/Lysmarine files and services installed, it makes sense to apply locale, timezone, keyboard layout and especially do security relevant things like SSH host key regeneration and machine ID reset.
And network setup according to dietp.txt choices.
Actually I don't see any step that could/should affect a pre-installed software.

Yep, The fact that it can be re-run without breaking anythig is a proof of good design. 
I now understand for the security stuff the need to run it even if it been pre-installed.

> As said only the dietpi-software prompt on first login could confuse or be skipped when having the desired set pre-installed 

Yeah that's one of my big consern, A fast, question-less and reboot-less first boot is a major point.  My userbase is not techy at all. Their expectation is you plug it and you can move a mouse 20sec later. I had to sell pre-installed SDcard to people overwhelmed by RuFus and etcher...

> we could simply solve this via /DietPi/dietpi/.skip_firstrun_installs 

That kind of mecanism could effectively do the job, I mean it's the goal. But idk about creating an other dot file. Would it be possible to use DIETPI_INSTALL_STAGE   ?

MichaIng replied:
@FredericGuilbault 
> I had to sell pre-installed SDcard to people overwhelmed by RuFus and etcher...

Ui, in such cases I am wondering if running an on SBC is such a good idea 😉. However learning by doing, was the same for me when I started with an RPi and ownCloud that day.

> Would it be possible to use DIETPI_INSTALL_STAGE ?

Currently I don't know how this could work fine. DietPi-Preboot (now DietPi-Firstboot) set DIETPI_INSTALL_STAGE=0, the first update =1, then it's always the same when DietPi-Software is reached. E.g. using a different start value means a more complicated handling withing this scripts. Since currently the only difference is indeed the DietPi-Software prompt and as of discussion it makes sense to keep difference as small as possible, an additional flag file is the cleanest method. It can be simply removed after it did it's job by the script:
```
if [[ -f /DietPi/dietpi/.skip_firstrun_installs ]]; then
    rm -f /{DietPi,boot}/dietpi/.skip_firstrun_installs
else
    Promt DietPi-Software for firstrun installs
fi
```

MichaIng replied:
@FredericGuilbault @Fourdee 
I have a simple solution: `/DietPi/dietpi/.installed`
- This file is created on first `dietpi-software` run, so it actually serves as a flag file already.
- All current pre-installed images ship software that is selectable in `dietpi-software`, so this file needs to be pre-created.
- Since actually all other firstrun setup steps make sense for pre-installed images as well, including the ones currently done in first `dietpi-software` execution, we could simply adjust the login prompt a bid, to make it clearer to users what to do e.g.:
```
if [[ $G_DIETPI_INSTALL_STAGE == 1 ]]; then
    prompt_text='Please select software to install, or, select "Install" to finish with a minimal system.'
    [[ -f /DietPi/dietpi/.installed ]] && prompt_text="Your DietPi image has the following software titles pre-installed: <show list, excluding Dropbear and DietPi-RAMlog, which are always pre-installed>"
    G_WHIP_MSG "$prompt_text"
    ...
fi
```
_________
We could even make the pre-installs selectable in DietPi-PREP already. The code is in place, network etc. is guaranteed, target device is known etc. Only architecture on RPi might be an issue, since some software titles are not available for ARMv6. So actually for RPi images this needs to be selected then as well. E.g. Allo GUI requires ARMv7.

But lets move the pre-install changes into a separate PR. I will not do some finish-up commits for firstrun separation.

MichaIng commented on dietpi/preboot:
```diff
@@ -299,12 +128,6 @@
 
 		fi
```
All the above as well should go into the new script, shouldn't it?

FredericGuilbault replied:
I tough those had to be run at every boot ?

MichaIng replied:
@FredericGuilbault 
Nope, `if (( $G_DIETPI_INSTALL_STAGE == -1 )); then` means fresh image, the firstrun step contains `echo 0 /DietPi/dietpi/.install_stage` which is read into `G_DIETPI_INSTALL_STAGE`.

It actually makes sense to put those into the same function. Most likely a coding relic that those are separate while all others are wrapped into a function.

MichaIng commented on rootfs/etc/systemd/system/dietpi-firstboot.service:
```diff
@@ -0,0 +1,14 @@
+Description=DietPi-First_boot_installation_process
+DefaultDependencies=false
+After= dietpi-preboot.service
+Before=dietpi-boot.service
```
`network-pre.target` should be added to `Before=` as well since all those steps are meant to run before network is being initialised (like it is for PreBoot).

FredericGuilbault replied:
Fixed with a18c393d04ea6e4e1e16024f546faaaa9545c136

MichaIng commented:
I just measured the execution time, to see if there is any effect of unused code:
```
root@VM-Stretch:/tmp# time for i in {1..10}; do ./preboot; done
...
real    0m11.830s
user    0m0.868s
sys     0m0.312s
root@VM-Stretch:/tmp# time for i in {1..10}; do /DietPi/dietpi/preboot; done
...
real    0m11.784s
user    0m0.924s
sys     0m0.212s
```
- So inerte code has no influence. I removed the whole install stage == -1 part + RPi_CPU_Clocks as well for this, as for completeness this should be done as well. As you can see no performance benefit (./preboot the one from this PR with above further reduction).
- However to have the boot script code cleaner and save some RAMdisk space should be reason enough.

Largest impact is of course the two child scripts (CPU governor + LED settings) that are executed in subshell and load dietpi-globals in case. Actually it should be at least possible to skip led settings, an additional setting to revert to system defaults, which removes the settings file and then scripts the execution. Actually we could move this as well as the CPU governor adjustment into an own systemd unit each. It does not really matter when they run, just after DietPi-RAMdisk. So they can be Type=simple to not delay any other boot scripts. This also makes it again simpler to revert to system defaults => disable the related systemd unit. Once applied settings (settings file) can be preserved then while toggling the service.

Finally with this the preboot script can be removed completely. Only thing that we need to run somewhere is:
- `/DietPi/dietpi/func/dietpi-obtain_hw_model`
- `(( $G_HW_MODEL < 10 )) && command -v amixer &> /dev/null && amixer set PCM -- -010`

Both are only required for RPi, hw_model estimation additionally is a failsafe thing for other devices as well.
- hw_model, if it runs at boot for all devices, it makes sense to run it as early as possible. Not beautiful but possible would be to merge it into dietpi-ramdisk 🤔. At least there it runs at the earliest possible point.
- amixer, I do not really understand what this is for @Fourdee ? Why volume needs adjustment and why this cannot be done e.g. via asound.conf or such?
However it should be possible to move this from preboot to boot script.

FredericGuilbault committed:
Typos and styling corrections .

MichaIng commented on rootfs/etc/systemd/system/dietpi-firstboot.service:
```diff
@@ -9,3 +9,6 @@ Type=oneshot
 ExecStart=/bin/dash -c '/DietPi/dietpi/dietpi-firstboot 2>&1 | tee /var/tmp/dietpi/logs/first_boot_install.log'
+
+[Install]
+WantedBy=local-fs.target
```
Tiny one: At best use multi-user.target here. The reason is that local-fs.target also runs at some rescue mode and we don't want to have any DietPi scripts running when it is known that login is never reached. Note that it doesn't matter for boot order or something which target you use, as long as the target is planned to be reached by systemd.

I also switched all other DietPi boot services to multi-user.target 1-2 versions ago to be consistent here, so nothing runs aside from base systemd in case something prevents multi-user.target (login prompts).

FredericGuilbault replied:
I didn't know about that. Thx

Fixed in a18c393d04ea6e4e1e16024f546faaaa9545c136

Fourdee commented:
@MichaIng 

> amixer, I do not really understand what this is for @Fourdee ? Why volume needs adjustment and why this cannot be done e.g. via asound.conf or such?
However it should be possible to move this from preboot to boot script.

This is resolved with current kernel under stretch. Can be removed.
It was previously needed as the PCM volume would reset (AFAIK, was a year+ ago lol) to lower volume after reboot. Not any more :+1: 

Done :u6307: https://github.com/MichaIng/DietPi/pull/2833#issuecomment-494811992

Fourdee committed:
No longer required.

https://github.com/MichaIng/DietPi/pull/2833#issuecomment-494811992

FredericGuilbault committed:
Move firstboot srcipt to /var/lib/dietpi/services/ as discussed

FredericGuilbault reviewed (commented):
In the commit 7760a2d904f2ddc965a9de8a05044feb5cf0dd43   The firstboot script have move to /var/lib/dietpi/services/   

I think this break @Fourdee   commit c0aa1e8cc044b8328f93bd53fffba6515e82769e

FredericGuilbault committed:
Add network-pre.target to firstboot.service

MichaIng committed:
v6.25

+ DietPi-Update | Revert firstboot script removal. This has now been moved to: /var/lib/dietpi/services/

MichaIng commented:
@FredericGuilbault 
Is it okay if we do commits to your branch? Usually I don't like it but it is possible, when a PR to our branch is opened. Thus I'm asking 😉.

FredericGuilbault commented:
Sure, no problem. I think you could also create a branch and modify this pullrequest to change it to the new branch and merge it.    But I don't mind it's up to you.

MichaIng committed:
v6.25

+ DietPi-PREP | Spelling

MichaIng committed:
v6.25

+ DietPi-Boot | Re-add swapfile creation on pre-installed images
+ DietPi-Boot | Failsafe network interface index estimation

MichaIng committed:
v6.25

+ DietPi-FirstBoot | Make dietpi-preboot.service a hard dependency, install to multi-user.target like all other boot scripts as well
+ DietPi-FirstBoot | Renamed to ".bash" file ending to make requirement clear, log to /tmp and leave moving to disk to DietPi-PostBoot

MichaIng committed:
v6.25

+ DietPi-FirstBoot | Add all "(( $G_DIETPI_INSTALL_STAGE == -1 ))" steps here
+ DietPi-FirstBoot | Rename to ".bash" file ending to make shell requirement clear
+ DietPi-FirstBoot | Do obtain network details here, which is already done in DietPi-PreBoot
+ DietPi-FirstBoot | Failsafe dietpi.txt and .network value grabbing, to avoid errors and instead apply defaults
+ DietPi-FirstBoot | Minor coding and alignment

MichaIng committed:
v6.25

+ DietPi-PreBoot | All firstrun setup steps have been moved to DietPi-FirstBoot

MichaIng committed:
v6.25

+ DietPi-PREP | Minor

MichaIng committed:
v6.25

+ DietPi-Boot | In attempt to resolve conflict due to: https://github.com/MichaIng/DietPi/commit/64bd543eff773e1fe60835d011b9e40934c36faa#diff-620191171f47b4763a532e984acd2e8b

MichaIng commented:
@FredericGuilbault @Fourdee 
<s>Hmm, I am not able to resolve the conflict. The web editor is simply loading forever. It is only due to this tiny change: https://github.com/MichaIng/DietPi/commit/64bd543eff773e1fe60835d011b9e40934c36faa#diff-620191171f47b4763a532e984acd2e8b

Will try it again later, but if you can do it, go for it, simply remove the "dev" part and fully use the version of this PR/branch, since I already committed the changes by Fourdee here as well: https://github.com/MichaIng/DietPi/pull/2833/commits/c28853cd2875159bf6130f0123329f3fc3a035ac</s>
#### EDIT: Made it, "HTTPS Everywhere" broke the web editor strangely 😄.

MichaIng committed:
Merge branch 'dev' into improvement/Seperate_first_boot

MichaIng committed:
v6.25

+ DietPi-PREP | DietPi-FirstBoot needs to be enabled on all systems

MichaIng committed:
v6.25

+ DietPi-FirstBoot | Tiny

MichaIng committed:
v6.25

+ CHANGELOG | DietPi-FirstBoot: First boot setup steps have been moved into a separate script and systemd unit

MichaIng committed:
v6.25

+ DietPi-FirstBoot | Fix RPi3 A+ default clocks

MichaIng commented:
Okay I merge the PR and run some tests.

FredericGuilbault commented:
Nice work :dancer:

Fourdee commented:
@FredericGuilbault 

Please email me an address to daniel.knight@dietpi.com. I'll ship some DietPi stickers out to you for your excellent work on this PR.

Apologies for the delay ^^
